cosmwasm/contracts/crosschain-swaps/src/execute.rs

```diff
@@ -1,24 +1,92 @@
 use cosmwasm_std::{coins, to_binary, wasm_execute, BankMsg, Env, Timestamp};
 use cosmwasm_std::{Addr, Coin, DepsMut, Response, SubMsg, SubMsgResponse, SubMsgResult};
-use registry::Registry;
+use registry::msg::{Callback, SerializableJson};
+use registry::{Registry, RegistryError};
 use swaprouter::msg::ExecuteMsg as SwapRouterExecute;
 
 use crate::checks::{check_is_contract_governor, ensure_key_missing, validate_receiver};
 use crate::consts::{MsgReplyID, CALLBACK_KEY};
-use crate::msg::{CrosschainSwapResponse, FailedDeliveryAction, SerializableJson};
+use crate::msg::{CrosschainSwapResponse, FailedDeliveryAction};
 use registry::proto::MsgTransferResponse;
 
-use crate::state;
 use crate::state::{
     Config, ForwardMsgReplyState, ForwardTo, SwapMsgReplyState, CONFIG, FORWARD_REPLY_STATE,
     INFLIGHT_PACKETS, RECOVERY_STATES, SWAP_REPLY_STATE,
 };
 use crate::utils::{build_memo, parse_swaprouter_reply};
 use crate::ContractError;
+use crate::{state, ExecuteMsg};
+
+/// This function takes any token. If it's already something we can work with
+/// (either native to osmosis or native to a chain connected to osmosis via a
+/// valid channel), it will just proceed to swap and forward. If it's not, then
+/// it will send an IBC message to unwrap it first and provide a callback to
+/// ensure the right swap_and_forward gets called after the unwrap succeeds
+pub fn unwrap_or_swap_and_forward(
+    deps: DepsMut,
+    block_time: Timestamp,
+    contract_addr: Addr,
+    swap_coin: Coin,
+    output_denom: String,
+    slippage: swaprouter::Slippage,
+    receiver: &str,
+    next_memo: Option<SerializableJson>,
+    failed_delivery_action: FailedDeliveryAction,
+) -> Result<Response, ContractError> {
+    deps.api
+        .debug(&format!("executing unwrap or swap and forward"));
+    let registry = Registry::default(deps.as_ref());
+    let path = registry.unwrap_denom_path(&swap_coin.denom)?;
+    if path.is_empty() {
+        return Err(RegistryError::InvalidDenomTracePath {
+            path: String::new(),
+            denom: swap_coin.denom,
+        }
+        .into());
+    }
+
+    // If the path is larger than 2, we need to unwrap this token first
+    if path.len() > 2 {
```

Comment:
So what precisely do we mean by step on `path` here? Is this a step in the registry, or denom trace length?

If its denom trace length, then we may run into issues with tokens that are natively 2 hops right. I know that there is one asset like this (pstake), due to some legacy reason.

We don't really need to have the contract handle cleanly swapping those assets, but it should fail "safely' in such scenarios.

Reply:
yeah, each element of the path is a hop (e.g. atom on juno sent to osmosis will have path length 3: osmosis, juno, hub). 

Didn't know 2-hop native assets was a thing. We could modify the registry so that we can tell which chain an asset is "native" to other than the default (native=not-ibc-wrapped), but for now, if we just don't include those in the registry, those calls will just fail with a path-not-found (or similar) error

Reply:
oh cool, path-not-found sounds like a safe failure! Can we write a go test to check that? If so, that sgtm

Reply:
(Can we add a comment here for definition of path length = IBC hops)